akka-stream/src/main/scala/akka/stream/ContextMapStrategy.scala

```diff
@@ -4,53 +4,35 @@
-   */
-  def same[In, Ctx, Out](): Iterate[In, Ctx, Out] = Iterate(
-    iterateFn = (_: In, inCtx: Ctx, _: Out, _: Int, _: Boolean) => inCtx
-  )
-
-  /**
-   * Transform the context of the first element given the input element [[In]] and context [[Ctx]].
-   */
-  def first[In, Ctx, Out](firstFn: (In, Ctx, Out) => Ctx): Iterate[In, Ctx, Out] = Iterate(
-    iterateFn = (in: In, inCtx: Ctx, out: Out, index: Int, _: Boolean) =>
-      if (index == 0) firstFn(in, inCtx, out)
-      else inCtx
-  )
+/**
+ * A ContextMapStrategy defines how the context parameter of a FlowWithContext is manipulated
+ * as various operations are performed in the flow
+ *
+ * @typeparam Ctx the context type
+ */
+class ContextMapStrategy[-Ctx]
 
+object ContextMapStrategy {
   /**
-   * Transform the context of the last element given the input element [[In]], context [[Ctx]], and the index
-   * of the element.
+   * Trait that is mixed into strategies that allow elements (including their offsets) to be filtered out of the stream.
    */
-  def last[In, Ctx, Out](fn: (In, Ctx, Out, Int) => Ctx): Iterate[In, Ctx, Out] = Iterate(
-    iterateFn = (in: In, inCtx: Ctx, out: Out, index: Int, hasNext: Boolean) =>
-      if (!hasNext) fn(in, inCtx, out, index)
-      else inCtx
-  )
+  trait Filtering[-Ctx] {
+    self: ContextMapStrategy[Ctx] =>
+  }
 
   /**
-   * Iterate over each element and transform the context given the element [[In]], context [[InCtx]],
-   * output element [[Out]], index, and if the there is a next output element [[Out]]. This strategy can be used to
-   * satisfy all 1:many use cases.
+   * Trait that is mixed into strategies that allow elements (including their offsets) to be reordered.
    */
-  def iterate[In, Ctx, Out](fn: (In, Ctx, Out, Int, Boolean) => Ctx): Iterate[In, Ctx, Out] = Iterate(fn)
+  trait Reordering[-Ctx] {
+    self: ContextMapStrategy[Ctx] =>
+  }
 
   /**
-   * Iterate over each element and transform the context given the element [[In]], context [[InCtx]],
-   * output element [[Out]], index, and if the there is a next output element [[Out]]. If there are no output elements
-   * then we can emit a single output element [[Out]] and output context [[Ctx]] that represents this state.
-   * This strategy can be used to satisfy all 1:many and 1:0 use cases.
+   * Trait that is mixed into strategies that allows operations like 'mapConcat', which
+   * turn a single element into zero or more elements, where it is known before the last
+   * element is emitted that this is the last element.
    */
-  def iterateOrEmpty[In, Ctx, Out](
-                                    fn: (In, Ctx, Out, Int, Boolean) => Ctx,
-                                    emptyFn: (In, Ctx) => (Out, Ctx)
-                                  ): Iterate[In, Ctx, Out] = Iterate(iterateFn = fn, emptyFn = Some(emptyFn))
+  trait Iteration[Ctx] extends Filtering[Ctx] { self: ContextMapStrategy[Ctx] =>
```

Comment:
Hmmm, interesting, `Iteration` _is a_ `Filtering`. Could this be the other way around?

From @seglo's comments and code, I got the impression we have 3 scenarios:

1. 1 incoming produces 1 (already supported using regular `map`)
2. 1 incoming produces 1 or many
3. 1 incoming produces 0 or 1

In the list above, `3.` is `iteration ` and `filtering` but, IIUC, `2.` is not `filtering` but is `iteration`. 

Then, you added `reordering` which is (sort of) a:

 - many incoming produces many (the same) outgoing

then, what happens if `mapAsyncUnordered` is preceded or succeeded by a `filter`?

Reply:
> Hmmm, interesting, `Iteration` _is a_ `Filtering`. Could this be the other way around?

Yes, but more precisely: 'a context mapping strategy that allows iteration _is a_ context mapping strategy that allows filtering'.

'filter' maps to '3', but 'mapConcat' does not map to '2' - it maps to '1 incoming produces 0-n'. I guess we could introduce a concept for '1 incoming produces 1 or many', but I'm not sure if we have any operators that actually map to that?

> then, what happens if `mapAsyncUnordered` is preceded or succeeded by a `filter`?

That is up to the strategy implementor. I actually have some ideas here, but indeed it would mean adding some complexity, so I wanted to keep it out of scope for the first iteration :).